Add biomappings for CTD ingest / Documentation updates

glass-ships opened the pull request:
Related to [Monarch Ingest #417](https://github.com/monarch-initiative/monarch-ingest/issues/417)

Pull in (with some alterations) [this SSSOM mapping file](https://github.com/biopragmatics/biomappings/blob/master/docs/_data/sssom/biomappings.sssom.tsv)

Also thought it might be nice to add in some more documentation for this repo, particularly about how the overall workflow, and how to add new mappings, like:
- Should an entry be made in the registry.yaml (and how to determine confidence level?) 
- Is it ok to edit the jinja template for the makefile? 
- What to edit in scripts
- etc.
- What is the "workflow" that the makefile generates?

matentzn commented on config/commons.Makefile.j2:
```diff
@@ -30,8 +30,12 @@ $(MAPPINGS_DIR)/{{ mapping.local_name }}: | $(MAPPINGS_DIR)/
-mappings: $(ALL_MAPPINGS)
+.PHONY: biomappings
+biomappings: | $(MAPPINGS_DIR)/
+	python3 scripts/process_biomappings.py
```
The commons.Makefile.j2 should not be edited this way:

YOu should

1. Add the mapping you want to create to https://github.com/monarch-initiative/monarch-mapping-commons/blob/main/registry.yaml
2. Run https://github.com/monarch-initiative/monarch-mapping-commons/blob/main/update_registry.sh to rebuild the makefile
3. Add the biomappings goals in https://github.com/monarch-initiative/monarch-mapping-commons/blob/main/custom.Makefile, which is the custom extension of the autogenerated Makefile.

matentzn commented on scripts/process_biomappings.py:
```diff
@@ -0,0 +1,19 @@
+df = df[(df['subject_id'].str.contains('chebi')) & (df['object_id'].str.contains('mesh'))]
+
+# Capitalize the mesh and chebi ids
+df['subject_id'] = df['subject_id'].str.upper()
```
We should avoid string hacking like this - it can easily go wrong in the future.

https://github.com/cthoyt/curies

```
converter.standardize_curie('gomf:0032571')
```

matentzn commented on scripts/process_biomappings.py:
```diff
@@ -0,0 +1,19 @@
+df['object_id'] = df['object_id'].str.upper()
+
+# Swap values of subject id and object id
+df['subject_id'], df['object_id'] = df['object_id'].values, df['subject_id'].values
```
We should use 

```
sssom invert ....
```

For this, because there is much more to flipping a mapping set then flipping the IDs.

Trivially, but not sufficiently, the subject_label and object_label columns must be flipped too, but there is more. Look at 

https://github.com/monarch-initiative/monarch-mapping-commons/blob/c12886f1c0e99740f0dc23bf09f095fb3224af12/mappings/biomappings.sssom.tsv#L1252

Flipping this the way you do will cause bigger errors!

glass-ships replied:
Thanks Nico! I think I've resolved the other two suggestions with this commit, but I'm having trouble with sssom invert... it looks like an invert mapping dictionary may be a required argument? but I'm not totally sure what that should look like

matentzn reviewed (changes requested):
@glass-ships I am glad we start working on this! For now, I will review all PRs on here until we get a shared understanding of the intended design :)

glass-ships commented:
Status (note to self):
`sssom.util.invert_mappings(...)` currently fails:
```
Traceback (most recent call last):
  File "/mnt/dev/dev/TISLab/monarch-mapping-commons/scripts/process_biomappings.py", line 39, in <module>
    df = sssom.util.invert_mappings(df=df, subject_prefix="MESH")
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/dev/dev/TISLab/monarch-mapping-commons/.venv/lib/python3.11/site-packages/sssom/util.py", line 1615, in invert_mappings
    [inverted_df, predicate_modified_df, non_inverted_df_by_predicate]
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnboundLocalError: cannot access local variable 'non_inverted_df_by_predicate' where it is not associated with a value
```

May require a `predicate_invert_dictionary`?

glass-ships commented:
> I'll make a release and updating the version of `sssom` in your project should let you proceed without issues.

Thanks so much for looking into this with me @hrshdhgd! This did resolve the error being thrown, but unfortunately I think I'm still missing something. I added in some assert statements to check what's going on, and it seems like the invert is somehow incomplete? 

assert statements start here: https://github.com/glass-ships/monarch-mapping-commons/blob/biomappings/scripts/process_biomappings.py#L38

<details><summary>Relevant output:</summary>

```
> python scripts/process_biomappings.py

      subject_id   subject_label     predicate_id     object_id    object_label         mapping_justification                  author_id  confidence mapping_tool
861  CHEBI:10001        Visnadin  skos:exactMatch  MESH:C067604        visnadin  semapv:ManualMappingCuration  orcid:0000-0001-9439-5346         NaN          NaN
862  CHEBI:10002        visnagin  skos:exactMatch  MESH:C044999        visnagin  semapv:ManualMappingCuration  orcid:0000-0001-9439-5346         NaN          NaN
863  CHEBI:10014       Voacamine  skos:exactMatch  MESH:C005891       voacamine  semapv:ManualMappingCuration  orcid:0000-0001-9439-5346         NaN          NaN
864  CHEBI:10025  Vulgaxanthin-I  skos:exactMatch  MESH:C018416  vulgaxanthin-I  semapv:ManualMappingCuration  orcid:0000-0001-9439-5346         NaN          NaN
865  CHEBI:10029        WIN54954  skos:exactMatch  MESH:C062554       Win 54954  semapv:ManualMappingCuration  orcid:0000-0001-9439-5346         NaN          NaN

      subject_id   subject_label     predicate_id     object_id    object_label         mapping_justification                  author_id  confidence mapping_tool
861  CHEBI:10001        Visnadin  skos:exactMatch  MESH:C067604        visnadin  semapv:ManualMappingCuration  orcid:0000-0001-9439-5346         NaN          NaN
862  CHEBI:10002        visnagin  skos:exactMatch  MESH:C044999        visnagin  semapv:ManualMappingCuration  orcid:0000-0001-9439-5346         NaN          NaN
863  CHEBI:10014       Voacamine  skos:exactMatch  MESH:C005891       voacamine  semapv:ManualMappingCuration  orcid:0000-0001-9439-5346         NaN          NaN
864  CHEBI:10025  Vulgaxanthin-I  skos:exactMatch  MESH:C018416  vulgaxanthin-I  semapv:ManualMappingCuration  orcid:0000-0001-9439-5346         NaN          NaN
865  CHEBI:10029        WIN54954  skos:exactMatch  MESH:C062554       Win 54954  semapv:ManualMappingCuration  orcid:0000-0001-9439-5346         NaN          NaN
Traceback (most recent call last):
  File "/mnt/dev/dev/TISLab/monarch-mapping-commons/scripts/process_biomappings.py", line 52, in <module>
    assert all(row.object_id.__contains__("CHEBI") for row in df.itertuples()), \
AssertionError: 
        Object IDs are not all CHEBI: ['CHEBI:10001' 'CHEBI:10002' 'CHEBI:10014' ... 'MESH:C420746'
 'MESH:C000511' 'MESH:C052875']
```
</details>

hrshdhgd commented:
@glass-ships , not a problem, my pleasure!

Note the param `merge_inverted` is Tru by default. What that means is the `inverted_df` is appended to the original df and hence your `assert` fails. If you notice, the resultant DataFrame's length is doubled. If you invert using:
```
df = sssom.util.invert_mappings(df=df, subject_prefix="MESH", merge_inverted=False)
```
Your asserts should pass.

glass-ships commented:
Ah i see! that did it thank you! 

i believe the biomapping specific changes are ready, although i did run into trouble when running `make all` to test, vs just running the script manually. (this may have not been a good idea, though)

matentzn commented on custom.Makefile:
```diff
@@ -17,5 +24,10 @@ MONDO_JSON=http://purl.obolibrary.org/obo/mondo/mondo.json
-	sssom parse $< --input-format obographs-json -m config/metadata.yaml --prefix-map-mode merged -o $@
+.PHONY: mondo_all
+mondo_all: $(TMP_DIR)/mondo.json
+	sssom parse $< --input-format obographs-json -m config/metadata.yaml --prefix-map-mode merged -o $(MAPPINGS_DIR)/mondo_all.sssom.tsv
```
This is a bit ideosyncractic for makefiles - we should only use phony goals to give names to specific pipelines. Why did you remove `$(MAPPINGS_DIR)/mondo_all.sssom.tsv`?

glass-ships replied:
oh i just forgot to switch it back, i think this was an attempt to fix an error in trying to run `make all`, I'm just not super fluent in a lot of the intricacies of makefiles

matentzn commented on pyproject.toml:
```diff
@@ -0,0 +1,22 @@
+description = "Mapping Commons for the Monarch Initiative knowledge graph"
+authors = [
+    "The Monarch Initiative <info@monarchinitiative.org>",
+    "Nico Matentzoglu <>",
```
```suggestion
    "Nico Matentzoglu <nicolas.matentzoglu@gmail.com>",
    "Harshad Hegde <hhegde@lbl.gov>",
    "Kevin Schaper <kevin@tislab.org>",
```

glass-ships commented on custom.Makefile:
```diff
@@ -17,5 +24,10 @@ MONDO_JSON=http://purl.obolibrary.org/obo/mondo/mondo.json
+
+.PHONY: biomappings
+biomappings: install
+	$(RUN) python3 scripts/biomappings.py $(MAPPINGS_DIR)/biomappings.sssom.tsv
```
interesting, is that just to avoid re-installing if the install "target" was already built?

matentzn replied:
Yes, it is only run once; but its always better to use non-phony goals as dependencies to really make sure of this.

tbh I wouldn't at this to the dependencies of a specific goal. I would simply add this to the SOP:

```
make install all
```

glass-ships commented:
A quick question about usage: 

I see the `run_build.sh` script builds the docker image, then runs `make all` and `make mappings` (which feels a bit redundant since both targets have the same directives?)

but then in `run.sh` the comment reads:
> Instead of typing "make TARGET", type "./run.sh make TARGET".

Does this mean the documentation should instead read:
> You can run a specific project like this:
```bash
./run.sh make symbiont-mondo-icd10cm

# rather than just 
make symbiont-mondo-icd10cm
```

hrshdhgd commented:
> Does this mean the documentation should instead read:
> 
> > You can run a specific project like this:
> 
> ```shell
> ./run.sh make symbiont-mondo-icd10cm
> 
> # rather than just 
> make symbiont-mondo-icd10cm
> ```

Yes.

glass-ships commented:
I've added a basic github action and python test which makes sure that the URL specified for each mapping item in the registry is live and valid, as it looks like the reason `make mappings` is failing is some dead references:
```python
tests/test_registry.py:32: TypeError
------ Captured stdout call ----------------------------------------
The following urls are not available:
http://w3id.org/sssom/commons/monarch/mondo_all.sssom.tsv
http://w3id.org/sssom/commons/monarch/upheno-all.sssom.tsv
http://w3id.org/sssom/commons/monarch/upheno-curated.sssom.tsv
http://w3id.org/sssom/commons/monarch/upheno-curated-negative.sssom.tsv
http://w3id.org/sssom/commons/monarch/upheno-lexical-oak.sssom.tsv
http://w3id.org/sssom/commons/monarch/upheno-lexical-custom.sssom.tsv
http://w3id.org/sssom/commons/monarch/upheno-logical-safe.sssom.tsv
```

glass-ships commented:
@hrshdhgd do you have any idea where these might be mirrored from? I checked [mondo](https://github.com/monarch-initiative/mondo/tree/master/src/ontology/mappings) but they don't appear to be there, but I think adding a `mirror_from` link for the missing URLs should resolve this? 

(or removing the lines which try to `wget` something from nonexistent w3id urls)

matentzn commented:
The URI prefix http://w3id.org/sssom/commons/monarch/ tells us that these mappings live in the Monarch mapping commons - some of these you will find in the mappings directory.

Feel free to remove all the ones that are not there or don't work for now, and open an issue to add them back..

glass-ships commented:
Sounds good, I'll go a head and do that for now. 

(So, do I correctly understand that monarch mapping commons is downloading those particular files from itself?)

glass-ships commented:
OK, i managed to get things running! 
Made the following changes:
- replaced deprecated `read_sssom_table` with `parse_sssom_table`
- added obographviz in docker container for og2dot commands to work
- commented out mapping sets with `http://w3id.org/sssom/commons/monarch/` prefix until we can get them working again

I ran `sh run.sh make mappings` to test things out, and after a few hours I ended up with over 10k files changed, so i opted to not commit them just yet

matentzn commented on scripts/process_biomappings.py:
```diff
@@ -0,0 +1,48 @@
+try:
+    biomappings_file = sys.argv[1]
+except IndexError:
+    biomappings_file = "https://raw.githubusercontent.com/biopragmatics/biomappings/master/docs/_data/sssom/biomappings.sssom.tsv"
```
Isnt there a purl for biomappings? I don't really like this kind of logic here, burying URLs into a script. I would remove the try catch and enforce the parameter being passed in.

glass-ships replied:
I don't think there are any purls for biopragmatics mapping files (see https://github.com/biopragmatics/biomappings/ where they offer the repo itself as the source of the files) 

But I'll add in an enforced argument for input file, should be easy enough, likewise with the output name

matentzn replied:
Great.

matentzn commented on scripts/process_biomappings.py:
```diff
@@ -0,0 +1,48 @@
+    f"\n\tSubject IDs are not all MESH: {df.subject_id.unique()}\n"
+
+# Write to file
+df.to_csv("mappings/mesh-chebi.biomappings.sssom.tsv", sep='\t', index=False)
```
"mappings/mesh-chebi.biomappings.sssom.tsv" should be a parameter to the script, in case the file location changes.

matentzn reviewed (commented):
I made some comments but it looks great to me!

glass-ships commented on scripts/process_biomappings.py:
```diff
@@ -0,0 +1,48 @@
+        prefix_synonyms=["mesh"],
+        uri_prefix="http://purl.obolibrary.org/obo/MESH_",
+    )
+])
```
How do you mean externalize? 

Unfortunately, for some reason the `get_monarch_converter()` method doesn't work and results in a blank output file

matentzn replied:
Just add a Json context file into the repo root directory for now and parse that, I can change the source of the map later. In the interest of moving on, if to hard, make an issue to replace the in line prefix maps with a proper Json LD context.

glass-ships commented:
@matentzn just wanted to check in, how's this looking?

matentzn reviewed (approved):
Looks good to make a start!

glass-ships commented:
so I just realized I don't have write access - if this is ready would you be willing to pull it in @matentzn?
